mkosi/cage.py

```diff
@@ -0,0 +1,669 @@
+    capabilities are lost again when the user executes a subprocess. As we also want subprocesses invoked by the user
+    to be able to mount stuff, we make sure the capabilities are inherited by adding all the user's capabilities to the
+    inherited and ambient capabilities set, which makes sure that they are passed down to subprocesses.
+    """
+    header = cap_user_header_t(LINUX_CAPABILITY_VERSION_3, 0)
+    payload = (cap_user_data_t * LINUX_CAPABILITY_U32S_3)()
+
+    if libc.capget(ctypes.addressof(header), ctypes.byref(payload)) < 0:
+        oserror()
+
+    payload[0].inheritable = payload[0].permitted
+    payload[1].inheritable = payload[1].permitted
+
+    if libc.capset(ctypes.addressof(header), ctypes.byref(payload)) < 0:
+        oserror()
+
+    effective = payload[1].effective << 32 | payload[0].effective
+
+    with open("/proc/sys/kernel/cap_last_cap", "rb") as f:
+        last_cap = int(f.read())
+
+    libc.prctl.argtypes = (ctypes.c_int, ctypes.c_ulong, ctypes.c_ulong, ctypes.c_ulong, ctypes.c_ulong)
+
+    for cap in range(ctypes.sizeof(ctypes.c_uint64) * 8):
+        if cap > last_cap:
+            break
+
+        if effective & (1 << cap) and libc.prctl(PR_CAP_AMBIENT, PR_CAP_AMBIENT_RAISE, cap, 0, 0) < 0:
+            oserror()
+
+
+def have_effective_cap(capability: int) -> bool:
+    with open("/proc/self/status") as f:
+        for line in f.readlines():
+            if line.startswith("CapEff:"):
+                hexcap = line.removeprefix("CapEff:").strip()
+                break
+        else:
+            return False
+
+    return (int(hexcap, 16) & (1 << capability)) != 0
+
+
+def seccomp_suppress_chown() -> None:
+    """
+    There's still a few files and directories left in distributions in /usr and /etc that are not owned by root. This
+    causes package managers to fail to install the corresponding packages when run from a single uid user namespace.
+    Unfortunately, non-root users can only create files owned by their own uid. To still allow non-root users to build
+    images, if requested we install a seccomp filter that makes calls to chown() and friends a noop.
+    """
+    libseccomp = ctypes.CDLL("libseccomp.so.2")
+
+    libseccomp.seccomp_init.argtypes = (ctypes.c_uint32,)
+    libseccomp.seccomp_init.restype = ctypes.c_void_p
+    libseccomp.seccomp_release.argtypes = (ctypes.c_void_p,)
+    libseccomp.seccomp_release.restype = None
+    libseccomp.seccomp_syscall_resolve_name.argtypes = (ctypes.c_char_p,)
+    libseccomp.seccomp_rule_add_exact.argtypes = (ctypes.c_void_p, ctypes.c_uint32, ctypes.c_int, ctypes.c_uint)
+    libseccomp.seccomp_load.argtypes = (ctypes.c_void_p,)
+
+    seccomp = libseccomp.seccomp_init(SCMP_ACT_ALLOW)
+
+    try:
+        for syscall in ("chown", "chown32", "fchown", "fchown32", "fchownat", "lchown", "lchown32"):
+            id = libseccomp.seccomp_syscall_resolve_name(syscall.encode())
```

Comment:
Nit: sincce it's all static we can avoid the encode.
```suggestion
        for syscall in (b"chown", b"chown32", b"fchown", b"fchown32", b"fchownat", b"lchown", b"lchown32"):
            id = libseccomp.seccomp_syscall_resolve_name(syscall)
```